Add torch.library.register_kernel

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124414
* #124403
* #124199
* #124134
* __->__ #124299
* #124200
* #124180

This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

zou3519 committed:
Add torch.library.register_kernel

This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_kernel"


This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_kernel"


This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_kernel"


This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_kernel"


This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on docs/source/library.rst:
```diff
@@ -21,12 +21,12 @@ Use :func:`torch.library.custom_op` to create new custom ops.
 
-Use the impl methods, such as :func:`torch.library.impl` and
-func:`torch.library.impl_abstract`, to add implementations
+Use the register methods, such as :func:`torch.library.register_kernel` and
```
```suggestion
Use the register_* methods, such as :func:`torch.library.register_kernel` and
```
Maybe?

albanD commented on docs/source/library.rst:
```diff
@@ -53,3 +53,5 @@ A tutorial that walks you through some examples on how to use this API is availa
 
 .. autofunction:: define
+
+.. autofunction:: impl
```
Agreed that moving this one here is the right move!

albanD commented on torch/library.py:
```diff
@@ -435,6 +435,65 @@ def impl_abstract(qualname, func=None, *, lib=None, _stacklevel=1):
+            the given device types.
+        device_types (None | str | Sequence[str]): The device_types to register an impl to.
+            If None, we will register to all device types -- please only use
+            this option if your implementation is truly device-type-agnostic.
```
Should we give a good example here? Maybe a view function that doesn't touch any data?

albanD reviewed (approved):
SGTM!

zou3519 committed:
Update on "Add torch.library.register_kernel"


This mirrors the .register_kernel method on the object produced by the
custom_op decorator.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "flaky"
